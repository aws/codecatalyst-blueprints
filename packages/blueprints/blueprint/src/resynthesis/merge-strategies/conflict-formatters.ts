import { CONFLICT_MARKER_LENGTH, splitLines } from './diff3';

export type MergeConflictFormatter = (a: string, o: string, b: string, options?: MergeConflictFormatterOptions) => string;

export interface MergeConflictFormatterOptions {
  aLabel?: string;
  oLabel?: string;
  bLabel?: string;
}

export class ConflictFormatters {
  public static diff3(a: string, o: string, b: string, options?: MergeConflictFormatterOptions): string {
    return createConflictString(a, b, {
      oText: o,
      aLabel: options?.aLabel,
      oLabel: options?.oLabel,
      bLabel: options?.bLabel,
    });
  }

  public static diff3NoAncestor(a: string, _o: string, b: string, options?: MergeConflictFormatterOptions) {
    return createConflictString(a, b, {
      aLabel: options?.aLabel,
      bLabel: options?.bLabel,
    });
  }

  public static preferExisting(a: string, _o: string, _b: string): string {
    return a;
  }

  public static preferProposed(_a: string, _o: string, b: string): string {
    return b;
  }

  public static trimEnds(a: string, _o: string, b: string, options?: MergeConflictFormatterOptions): string {
    const aLines = splitLines(a);
    const bLines = splitLines(b);

    const resultLines: string[] = [];
    const minLength = Math.min(aLines.length, bLines.length);
    for (let i = 0; i < minLength; i++) {
      if (aLines[i] !== bLines[i]) {
        break;
      }

      resultLines.push(aLines[i]);
    }

    const commonPrefixLength = resultLines.length;
    let commonSuffix: string[] = [];
    for (let i = 0; i < minLength - commonPrefixLength; i++) {
      if (aLines[aLines.length - 1 - i] !== bLines[bLines.length - 1 - i]) {
        break;
      }

      commonSuffix.push(aLines[aLines.length - 1 - i]);
    }

    const aConflictLines = aLines.slice(commonPrefixLength, aLines.length - commonSuffix.length);
    const bConflictLines = bLines.slice(commonPrefixLength, bLines.length - commonSuffix.length);
    if (aConflictLines.length || bConflictLines.length) {
      resultLines.push(createConflictString(aConflictLines.join(''), bConflictLines.join(''), { ...options }));
    }

    if (commonSuffix.length) {
      resultLines.push(...commonSuffix.reverse());
    }

    return resultLines.join('');
  }
}

function createConflictString(
  aText: string,
  bText: string,
  options?: {
    aLabel?: string;
    bLabel?: string;
    oText?: string;
    oLabel?: string;
  },
) {
  let chunkContents = '';

  chunkContents += createConflictMarker('<', options?.aLabel);
  chunkContents += aText;
  if (!aText.endsWith('\n')) {
    chunkContents += '\n';
  }

  if (options?.oText !== undefined) {
    chunkContents += createConflictMarker('|', options?.oLabel);
    chunkContents += options.oText;
    if (!options.oText.endsWith('\n')) {
      chunkContents += '\n';
    }
  }

  chunkContents += createConflictMarker('=');
  chunkContents += bText;
  if (!bText.endsWith('\n')) {
    chunkContents += '\n';
  }
  chunkContents += createConflictMarker('>', options?.bLabel);

  return chunkContents;
}

function createConflictMarker(type: '<' | '|' | '=' | '>', label?: string): string {
  return type.repeat(CONFLICT_MARKER_LENGTH) + `${label ? ' ' + label : ''}\n`;
}
