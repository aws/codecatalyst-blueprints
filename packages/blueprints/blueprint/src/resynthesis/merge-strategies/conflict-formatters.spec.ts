import { ConflictFormatters } from './conflict-formatters';

[
  ConflictFormatters.diff3,
  ConflictFormatters.diff3NoAncestor,
  ConflictFormatters.preferExisting,
  ConflictFormatters.preferProposed,
  ConflictFormatters.trimEnds,
].forEach(formatter => {
  describe(`conflict formatter: ${formatter.name}`, () => {
    it('formats simple conflicts', () => {
      const aText = 'line 1 a\nline 2';
      const oText = 'line 1 o\nline 2';
      const bText = 'line 1 b\nline 2';

      expect(formatter(aText, oText, bText, { aLabel: 'existing', oLabel: 'ancestor', bLabel: 'proposed' })).toMatchSnapshot();
    });

    it('formats simple conflicts with trailing newlines', () => {
      const aText = 'line 1 a\nline 2\n';
      const oText = 'line 1 o\nline 2\n';
      const bText = 'line 1 b\nline 2\n';

      expect(formatter(aText, oText, bText, { aLabel: 'existing', oLabel: 'ancestor', bLabel: 'proposed' })).toMatchSnapshot();
    });

    it('formats a conflict with a common prefix and suffix', () => {
      const commonPrefix = 'common prefix line\n'.repeat(3);
      const commonSuffix = 'common suffix line\n'.repeat(3);
      const aText = `${commonPrefix}some contents from a\n${commonSuffix}`;
      const oText = `${commonPrefix}some contents from o\n${commonSuffix}`;
      const bText = `${commonPrefix}some contents from b\n${commonSuffix}`;

      expect(formatter(aText, oText, bText, { aLabel: 'existing', oLabel: 'ancestor', bLabel: 'proposed' })).toMatchSnapshot();
    });
  });
});
