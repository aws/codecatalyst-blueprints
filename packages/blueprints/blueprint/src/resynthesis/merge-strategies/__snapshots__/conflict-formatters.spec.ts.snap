// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conflict formatter: diff3 formats a conflict with a common prefix and suffix 1`] = `
"<<<<<<< existing
common prefix line
common prefix line
common prefix line
some contents from a
common suffix line
common suffix line
common suffix line
||||||| ancestor
common prefix line
common prefix line
common prefix line
some contents from o
common suffix line
common suffix line
common suffix line
=======
common prefix line
common prefix line
common prefix line
some contents from b
common suffix line
common suffix line
common suffix line
>>>>>>> proposed
"
`;

exports[`conflict formatter: diff3 formats simple conflicts 1`] = `
"<<<<<<< existing
line 1 a
line 2
||||||| ancestor
line 1 o
line 2
=======
line 1 b
line 2
>>>>>>> proposed
"
`;

exports[`conflict formatter: diff3 formats simple conflicts with trailing newlines 1`] = `
"<<<<<<< existing
line 1 a
line 2
||||||| ancestor
line 1 o
line 2
=======
line 1 b
line 2
>>>>>>> proposed
"
`;

exports[`conflict formatter: diff3NoAncestor formats a conflict with a common prefix and suffix 1`] = `
"<<<<<<< existing
common prefix line
common prefix line
common prefix line
some contents from a
common suffix line
common suffix line
common suffix line
=======
common prefix line
common prefix line
common prefix line
some contents from b
common suffix line
common suffix line
common suffix line
>>>>>>> proposed
"
`;

exports[`conflict formatter: diff3NoAncestor formats simple conflicts 1`] = `
"<<<<<<< existing
line 1 a
line 2
=======
line 1 b
line 2
>>>>>>> proposed
"
`;

exports[`conflict formatter: diff3NoAncestor formats simple conflicts with trailing newlines 1`] = `
"<<<<<<< existing
line 1 a
line 2
=======
line 1 b
line 2
>>>>>>> proposed
"
`;

exports[`conflict formatter: preferExisting formats a conflict with a common prefix and suffix 1`] = `
"common prefix line
common prefix line
common prefix line
some contents from a
common suffix line
common suffix line
common suffix line
"
`;

exports[`conflict formatter: preferExisting formats simple conflicts 1`] = `
"line 1 a
line 2"
`;

exports[`conflict formatter: preferExisting formats simple conflicts with trailing newlines 1`] = `
"line 1 a
line 2
"
`;

exports[`conflict formatter: preferProposed formats a conflict with a common prefix and suffix 1`] = `
"common prefix line
common prefix line
common prefix line
some contents from b
common suffix line
common suffix line
common suffix line
"
`;

exports[`conflict formatter: preferProposed formats simple conflicts 1`] = `
"line 1 b
line 2"
`;

exports[`conflict formatter: preferProposed formats simple conflicts with trailing newlines 1`] = `
"line 1 b
line 2
"
`;

exports[`conflict formatter: trimEnds formats a conflict with a common prefix and suffix 1`] = `
"common prefix line
common prefix line
common prefix line
<<<<<<< existing
some contents from a
=======
some contents from b
>>>>>>> proposed
common suffix line
common suffix line
common suffix line
"
`;

exports[`conflict formatter: trimEnds formats simple conflicts 1`] = `
"<<<<<<< existing
line 1 a
=======
line 1 b
>>>>>>> proposed
line 2"
`;

exports[`conflict formatter: trimEnds formats simple conflicts with trailing newlines 1`] = `
"<<<<<<< existing
line 1 a
=======
line 1 b
>>>>>>> proposed
line 2
"
`;
